src/OrasProject.Oras/Extensions.cs

```diff
@@ -42,32 +68,88 @@ public static async Task<Descriptor> CopyAsync(this ITarget src, string srcRef,
+        {
+            (root, rootStream) = await srcRefFetchable.FetchAsync(srcRef, cancellationToken).ConfigureAwait(false);
+        } else
+        {
+            root = await src.ResolveAsync(srcRef, cancellationToken).ConfigureAwait(false);
+            rootStream = await src.FetchAsync(root, cancellationToken).ConfigureAwait(false);
+        }
+        var proxy = new Proxy()
+        {
+            MemoryStorage = new MemoryStorage(),
+            Target = src
+        };
+        if (Descriptor.IsManifestType(root))
+        {
+            if (!await proxy.MemoryStorage.ExistsAsync(root, cancellationToken).ConfigureAwait(false))
+            {
+                await proxy.MemoryStorage.PushAsync(root, rootStream, cancellationToken).ConfigureAwait(false);
+            }
+        }
+        await src.CopyGraphAsync(dst, root, proxy, cancellationToken).ConfigureAwait(false);
         await dst.TagAsync(root, dstRef, cancellationToken).ConfigureAwait(false);
         return root;
     }
 
-    public static async Task CopyGraphAsync(this ITarget src, ITarget dst, Descriptor node, CancellationToken cancellationToken)
+    /// <summary>
+    /// CopyGraphAsync concurrently copy node from src to dst by using proxy cache
+    /// </summary>
+    /// <param name="src"></param>
+    /// <param name="dst"></param>
+    /// <param name="node"></param>
+    /// <param name="proxy"></param>
+    /// <param name="cancellationToken"></param>
+    public static async Task CopyGraphAsync(this ITarget src, ITarget dst, Descriptor node, Proxy proxy, CancellationToken cancellationToken)
     {
-        // check if node exists in target
-        if (await dst.ExistsAsync(node, cancellationToken).ConfigureAwait(false))
+        // acquire lock to find successors of the current node
+        await _semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+        IEnumerable<Descriptor> successors;
+        try
+        {        
+            // check if node exists in target
+            if (await dst.ExistsAsync(node, cancellationToken).ConfigureAwait(false))
+            {
+                return;
+            }
+            successors = await proxy.GetSuccessorsAsync(node, cancellationToken).ConfigureAwait(false);
+        }
+        finally
         {
-            return;
+            _semaphore.Release();
         }
-
-        // retrieve successors
-        var successors = await src.GetSuccessorsAsync(node, cancellationToken).ConfigureAwait(false);
-        // obtain data stream
-        var dataStream = await src.FetchAsync(node, cancellationToken).ConfigureAwait(false);
-        // check if the node has successors
-        if (successors != null)
+        
+        var childNodesCopies = new List<Task>();
+        foreach (var childNode in successors)
+        {
+            childNodesCopies.Add(Task.Run(async () => await src.CopyGraphAsync(dst, childNode, proxy, cancellationToken).ConfigureAwait(false), cancellationToken));
```

Comment:
nit: How about formatting it like this for better readability:
```suggestion
                        childNodesCopies.Add(Task.Run(async () =>
                        await src.CopyGraphAsync(dst, childNode, proxy, cancellationToken).ConfigureAwait(false),
                        cancellationToken
                )
            );
```